File: week6/tcp_client.c
#include <stdio.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include "string.h"
#include "stdlib.h"
#include "unistd.h"
 
#define IP "127.0.0.1"
#define PORT 7777
#define BUFFER_SIZE 1024

int main()
{
    int client_socket = 0, ret = 0, len = 0;
    char buf[BUFFER_SIZE] = {0};
    struct sockaddr_in client_sockaddr;
    len = sizeof(client_sockaddr);
    //建立监听套接字
    client_socket = socket(AF_INET,SOCK_STREAM,0);
 
    //绑定IP地址和端口号
    client_sockaddr.sin_family = AF_INET;
    client_sockaddr.sin_port = htons(PORT);
    client_sockaddr.sin_addr.s_addr = inet_addr(IP);
    //建立链接
    ret = connect(client_socket,(struct sockaddr *)&client_sockaddr, len);
    if(ret == -1)
    {
        perror("connect");
        return -1;
    }
    printf("链接成功！\n");

    while(1)
    {
        //发送
        memset(buf, 0, BUFFER_SIZE);
        printf("发送的数据>>\n");
        fgets(buf, BUFFER_SIZE, stdin);
        //send(client_socket,buf,strlen(buf),0);
        write(client_socket, buf, strlen(buf));
        //接受数据
        memset(buf,0,BUFFER_SIZE);
        //recv(client_socket,buf,BUFFER_SIZE,0);
        read(client_socket,buf,BUFFER_SIZE);
        printf("收到的数据>>%s\n",buf);
        
    }
 
    close(client_socket);
}

File: week6/tcp_client.go
package main

import (
	"fmt"
	"net"
	"os"
)

func main() {
	//发起连接请求
	conn, err := net.Dial("tcp", "127.0.0.1:7777")
	if err != nil {
		fmt.Println("net.dial err:", err)
		return
	}
	defer conn.Close()
	//获取用户的键盘输入，将输入数据发送给服务器
	go func() {
		str := make([]byte, 4096)
		for {
			n, err := os.Stdin.Read(str)
			if err != nil {
				fmt.Println("os.Stdin.Read err:", err)
				continue
			}
			//写给服务器,读多少，写多少
			conn.Write(str[:n])
		}
	}()
	//打印服务器回发的大写数据
	buf := make([]byte, 4096)
	for {
		n, err := conn.Read(buf)
		if n == 0 {
			fmt.Println("检测到连接已断开，退出客户端")
			return
		}
		if err != nil {
			fmt.Println("conn.Read err:", err)
			return
		}
		fmt.Println("客户端收到服务器回复：", string(buf[:n]))
	}
}
